provider/bidengine/pricing.go

```diff
@@ -1,15 +1,123 @@
+	"errors"
+	"fmt"
+	"math/big"
+	"os/exec"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	"github.com/ovrclk/akash/types/unit"
 	"github.com/ovrclk/akash/validation"
 	dtypes "github.com/ovrclk/akash/x/deployment/types"
 )
 
-func calculatePrice(gspec *dtypes.GroupSpec) (sdk.Coin, error) {
+type BidPricingStrategy interface {
+	calculatePrice(gspec *dtypes.GroupSpec) (sdk.Coin, error)
+}
+
+const denom = "uakt"
+
+type scalePricing struct {
+	cpuScale     uint64
+	memoryScale  uint64
+	storageScale uint64
+}
+
+func MakeScalePricing(
+	cpuScale uint64,
+	memoryScale uint64,
+	storageScale uint64) (BidPricingStrategy, error) {
+
+	if cpuScale == 0 && memoryScale == 0 && storageScale == 0 {
+		return nil, fmt.Errorf("At least one bid price must be a non-zero number")
+	}
+
+	result := scalePricing{
+		cpuScale:     cpuScale,
+		memoryScale:  memoryScale,
+		storageScale: storageScale,
+	}
+
+	return result, nil
+}
+
+var ErrBidQuantityInvalid = errors.New("A bid quantity is invalid")
+var ErrBidZero = errors.New("A bid of zero was produced")
+
+func (fp scalePricing) calculatePrice(gspec *dtypes.GroupSpec) (sdk.Coin, error) {
+	// Use unlimited precision math here.
+	// Otherwise a correctly crafted order could create a cost of '1' given
+	// a possible configuration
+	cpuTotal := big.NewInt(0)
+	memoryTotal := big.NewInt(0)
+	storageTotal := big.NewInt(0)
+
+	// iterate over everything & sum it up
+	for _, group := range gspec.Resources {
+
+		groupCount := big.NewInt(0)
+		groupCount.SetUint64(uint64(group.Count)) // Expand uint32 to uint64
+
+		cpuQuantity := big.NewInt(0)
+		cpuQuantity.SetUint64(group.Resources.CPU.Units.Val.Uint64())
+		cpuQuantity.Mul(cpuQuantity, groupCount)
+		cpuTotal.Add(cpuTotal, cpuQuantity)
+
+		memoryQuantity := big.NewInt(0)
+		memoryQuantity.SetUint64(group.Resources.Memory.Quantity.Value())
+		memoryQuantity.Mul(memoryQuantity, groupCount)
+		memoryTotal.Add(memoryTotal, memoryQuantity)
+
+		storageQuantity := big.NewInt(0)
+		storageQuantity.SetUint64(group.Resources.Storage.Quantity.Val.Uint64())
+		storageQuantity.Mul(storageQuantity, groupCount)
+		storageTotal.Add(storageTotal, storageQuantity)
+	}
+
+	scale := big.NewInt(0)
+	scale.SetUint64(fp.cpuScale)
+	cpuTotal.Mul(cpuTotal, scale)
+
+	scale.SetUint64(fp.memoryScale)
+	memoryTotal.Mul(memoryTotal, scale)
+
+	scale.SetUint64(fp.storageScale)
+	storageTotal.Mul(storageTotal, scale)
+
+	// Each quantity must be non negative
+	// and fit into an Int64
+	if cpuTotal.Sign() < 0 || !cpuTotal.IsInt64() ||
+		memoryTotal.Sign() < 0 || !memoryTotal.IsInt64() ||
+		storageTotal.Sign() < 0 || !storageTotal.IsInt64() {
+		return sdk.Coin{}, ErrBidQuantityInvalid
+	}
+
+	cpuCost := sdk.NewCoin(denom, sdk.NewIntFromBigInt(cpuTotal))
+	memoryCost := sdk.NewCoin(denom, sdk.NewIntFromBigInt(memoryTotal))
+	storageCost := sdk.NewCoin(denom, sdk.NewIntFromBigInt(storageTotal))
+
+	// Check for less than or equal to zero
+	cost := cpuCost.Add(memoryCost).Add(storageCost)
```

Comment:
do the overflow checks after this.

Reply:
Doesn't the coin object use a big integer internally?

Reply:
ah, true.  too late for me.